substreams/ethereum-nabla/src/modules/map_components.rs

```diff
@@ -0,0 +1,195 @@
+
+    let name = abi::swap_pool::functions::Name {}
+        .call(pool_address.clone())
+        .ok_or_else(|| format!("Failed to retrieve 'name' for pool {:?}", pool_id))?;
+
+    let symbol = abi::swap_pool::functions::Symbol {}
+        .call(pool_address.clone())
+        .ok_or_else(|| format!("Failed to retrieve 'symbol' for pool {:?}", pool_id))?;
+
+    let attributes = vec![
+        Attribute {
+            name: "name".to_string(),
+            value: name.into(),
+            change: ChangeType::Creation.into(),
+        },
+        Attribute {
+            name: "symbol".to_string(),
+            value: symbol.into(),
+            change: ChangeType::Creation.into(),
+        },
+    ];
+
+    Ok(ProtocolComponent {
+        id: pool_id,
+        tokens: vec![asset],
+        tx: Some(tx.into()),
+        change: i32::from(ChangeType::Creation),
+        contracts: vec![],
+        static_att: attributes,
+        protocol_type: Option::from(ProtocolType {
+            name: ComponentType::SwapPool.into(),
+            financial_type: FinancialType::Swap.into(),
+            attribute_schema: vec![],
+            implementation_type: ImplementationType::Custom.into(),
+        }),
+    })
+}
+
+fn detect_asset_registered_event(log: &Log) -> Option<abi::nabla_portal::events::AssetRegistered> {
+    (log.address == NABLA_PORTAL.contract_address)
+        .then(|| abi::nabla_portal::events::AssetRegistered::match_and_decode(log))?
+}
+
+fn to_component(
+    tx: &TransactionTrace,
+    asset_registered: abi::nabla_portal::events::AssetRegistered,
+) -> Result<[ProtocolComponent; 2], String> {
+    let router_id = asset_registered.router.to_hex();
+    let asset_id = asset_registered.asset.to_hex();
+
+    let pool_address = abi::nabla_router::functions::PoolByAsset { param0: asset_registered.asset }
+        .call(asset_registered.router.clone())
+        .ok_or_else(|| {
+            format!(
+                "Failed to retrieve pool_address for asset {} from the router {}",
+                asset_id, router_id
+            )
+        })?;
+
+    let router = initialize_router(tx, asset_registered.router)?;
+    let swap_pool = initialize_swap_pool(tx, pool_address.clone())?;
+    Ok([router, swap_pool])
+}
+
+fn collect_components(
+    tx: &TransactionTrace,
+) -> Result<Option<TransactionProtocolComponents>, String> {
+    let components: Vec<ProtocolComponent>;
+    if tx.hash == NABLA_PORTAL.tx_hash {
+        components = vec![initialize_portal(tx)];
+    } else {
+        components = tx
+            .logs_with_calls()
+            .filter_map(|(log, _call)| detect_asset_registered_event(log))
+            .map(|asset_registered| to_component(tx, asset_registered))
+            .flatten_ok()
+            .try_collect()?;
+    }
+    Ok((!components.is_empty())
+        .then(|| TransactionProtocolComponents { tx: Some(tx.into()), components }))
+}
```

Comment:
when the deployment transaction for the portal is found, we know for a fact that no events can be emitted by the portal within the same block, hence `if` / `else` is fine here. `flatten_ok()` in conjunction with `try_collect` allows us to propagate the error (which should never occur, but other implementations would either require: 1. intermediate collect or 2. silencing error (using `.flatten().flatten()` would do that; first flatten would filter our Error (or None) and error would go unnoticed). This way the iterator remains lazy till the very end, while still short-circuiting in case of an error.